### Verification — `032bbe4` (rev 1)

| # | Claim | Result |
| --- | --- | --- |
| 1 | 가드가 루프 형태로 판정한다 — 무한 `while` + 파싱 가능한 `sleep` 은 차단, 100s 미만 `for`·배경 실행·비-Bash 는 통과 | PASS(live) |
| 2 | 차단 메시지의 `Reference:` 가 132바이트 리다이렉트 stub 이 아니라 spec.md 를 **절대경로**로 지목한다 | PASS(live) |
| 3 | N+1 번째 차단부터 read-gate 로 에스컬레이트하고, 이 spec 을 읽으면 해제된다 | PASS(mirror) |
| 4 | jq-config dedup 상태가 프로세스별로 직렬화되어 동시 실행에서 유실되지 않는다 | PASS(live) |
| 5 | shellcheck job 이 통과한다 — 이 PR 의 실패 체크 2개의 공통 원인 | PASS(live) |

<details><summary>Unverified — (a) read-gate 에스컬레이션은 실제 dispatcher 세션에서 end-to-end 로 관측하지 않음, (b) second-failure-advisory 변경은 스위트로만 확인</summary>
<br>

**(a)** spec 자체가 이 한계를 명시합니다 — 에스컬레이션은 dispatcher 가 쓰는 RICH ledger 행에 의존하고, 단독 `python3 impl.py` 호출은 COARSE 행만 남겨 `count_session_fires` 가 걸러냅니다. 따라서 직접 호출 프로브로는 구조적으로 관측할 수 없고, 행 3 은 스위트를 오라클로 삼은 `PASS(mirror)` 입니다. 실 세션 관측은 이 가드가 같은 세션에서 3회 차단되는 상황을 기다려야 하므로 머지 후 자연 발생 시 확인이 맞습니다.

**(b)** `second-failure-advisory` 의 상태 직렬화 변경은 전용 스위트 18건과 동시성 테스트로만 확인했고, 별도 라이브 프로브는 돌리지 않았습니다. 행 4 의 동시성 테스트가 같은 `_lib` 상태 경로를 실제 프로세스로 밟으므로 완전한 공백은 아니지만, 이 훅 자체의 end-to-end 는 아닙니다.

</details>

<details><summary>Evidence 1 — 가드에 실제 PreToolUse payload 를 먹여 네 가지 형태를 한 세션에서 판정</summary>
<br>

이 가드의 오라클은 "무엇을 폴링하는가" 가 아니라 **루프 형태**입니다. 그래서 프로브는 무한 `while`, 임계 미만 `for`, 같은 루프의 배경 실행, 비-Bash 도구를 나란히 넣습니다. 임계 미만과 배경 실행이 통과해야 차단이 형태 판정이라는 뜻이 되고, 그것이 이 행의 음성 대조입니다.

```
$ python3 probe_guard.py
session probe-70735d37

  unbounded while + sleep 20      -> exit=2  BLOCKS
  bounded for, 5x18s = 90s < 100  -> exit=0  passes
  same unbounded loop, background -> exit=0  passes
  non-Bash tool (Read)            -> exit=0  passes
```

스위트도 같은 결론을 폭넓게 고정합니다:

```
$ bash tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
  passed: 54  failed: 0
```

</details>

<details><summary>Evidence 2 — 차단 메시지가 실제로 내놓은 `Reference:` 줄을 그대로 읽음</summary>
<br>

이슈 #1012 의 핵심은 참조가 저장소 상대경로의 리다이렉트 stub 을 가리켜, praxis 체크아웃 밖에서 일하는 에이전트에게는 아무것도 resolve 되지 않았다는 점입니다. 위 프로브가 차단당한 그 호출의 stderr 에서 `Reference:` 줄을 그대로 집어냅니다.

```
$ python3 probe_guard.py 2>&1 | sed "s|$HOME|<HOME>|g"
reference the block message names:
  Reference: <HOME>/projects/praxis-wt/issue-1012-poll-guard-read-gate/hooks/preflight-gate/foreground-poll-loop-guard/spec.md
```

절대경로이고, `docs/hook/...` stub 이 아니라 spec.md 본체입니다. 홈 경로는 `$HOME` 를 환경변수로 둔 `sed` 로 치환했고 명령은 실제로 그렇게 실행했습니다.

</details>

<details><summary>Evidence 3 — read-gate 11개 케이스, 해제 조건과 fail-open 을 양방향으로</summary>
<br>

행 3 이 `PASS(mirror)` 인 이유는 Unverified (a) 에 있습니다. 스위트가 오라클인 만큼, 이 게이트가 *걸리는* 경우만이 아니라 *걸리지 않아야 하는* 경우까지 덮는지가 관건입니다 — 임계 미만 침묵, 읽은 뒤 해제, 다른 `.md` 로는 해제되지 않음, 참조 미해결 시 fail-open, session_id 부재 시 침묵, 무장 상태에서도 배경 실행은 통과.

```
$ bash tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh 2>&1 | grep -E '^ok +\[readgate'
ok    [readgate-reference-is-absolute-spec]
ok    [readgate-below-threshold-silent]
ok    [readgate-nth-block-demands-read]
ok    [readgate-released-after-read]
ok    [readgate-other-md-read-does-not-satisfy]
ok    [readgate-unresolvable-reference-fails-open]
ok    [readgate-no-session-id-silent]
ok    [readgate-armed-still-passes-background]
ok    [readgate-dispatcher-written-count-escalates]
```

`readgate-unresolvable-reference-fails-open` 이 특히 중요합니다 — 자기 spec 을 찾지 못하는 게이트가 아무도 만족시킬 수 없는 요구를 추가하면 교착이 됩니다.

</details>

<details><summary>Evidence 4 — 실제 프로세스를 띄우는 동시성 테스트</summary>
<br>

이 행은 목이 아니라 실행으로만 답이 나옵니다. 해당 테스트는 `subprocess.Popen` 으로 실제 프로세스를 동시에 띄워 상태 파일 경합을 만듭니다 — dedup 경합, staging 파일명이 프로세스별인지, 실패 시 unlink 되는지까지.

```
$ python3 -m pytest tests/test_hook_state_concurrency.py -q
........                                                                 [100%]
8 passed in 2.86s
```

같은 계열의 `second-failure-advisory` 스위트도 함께 돌렸습니다. 이 스위트는 실패 시에만 FAIL 줄을 찍으므로, `PASS: 18` 만 출력되고 FAIL 줄이 없는 것이 통과 형태입니다:

```
$ bash tests/hooks/postuse-correction/test_second_failure_advisory.sh | tail -1
PASS: 18
```

</details>

<details><summary>Evidence 5 — CI 워크플로의 shellcheck 호출을 그대로 재현</summary>
<br>

이 PR 의 실패 체크 2개는 원인이 하나였습니다 — 테스트의 `2>&1 >/dev/null` 를 shellcheck 가 뒤바뀐 리다이렉션(SC2069)으로 읽고, 그 job 이 blocking 이었습니다. 그래서 오라클은 로컬 관습이 아니라 `ci.yml` 의 그 줄이어야 합니다. `find` 표현식과 `--severity=warning` 까지 워크플로에서 그대로 옮겼습니다.

```
$ find . -name '*.sh' -not -path './.git/*' -print0 | xargs -0 shellcheck --severity=warning > "$SC" 2>&1 ; echo "shellcheck exit=$?"
shellcheck exit=0
$ wc -l < "$SC"
       0
```

`$SC` 는 워크트리를 더럽히지 않으려고 출력 경로를 담아 둔 환경변수이고, 명령은 실제로 그 형태로 실행했습니다.

exit status 를 함께 남긴 이유는 이 프로브가 성공 시 침묵하기 때문입니다 — 빈 출력만으로는 "지적 없음" 과 "실행되지 않음" 이 구별되지 않습니다.

</details>

<details><summary>History — rev 1, 최초 게시</summary>
<br>

이전 판본 없음. 대체된 근거도 없습니다.

</details>

--- a/hooks/postuse-correction/second-failure-advisory/impl.py
+++ b/hooks/postuse-correction/second-failure-advisory/impl.py
@@ -14,9 +14,22 @@ Only the repeated *failure* path is handled:
 - Missing `session_id` -> fail-open (no output, exit 0)
 - Successful tool calls -> no state update, no output
 - First failure for a `(tool_name, signature)` pair -> no output
-- Second failure for the same pair in one session -> stderr advisory
-- Third+ failure for the same pair -> no additional advisory by design (tracked by
-  the dedicated `second` count at `1 -> advisory` boundary)
+- Second failure for the same pair in one session -> advisory
+- Third+ failure for the same pair -> the advisory REPEATS, carrying the running
+  occurrence number (issue #1012)
+
+Why 3..N also advise (issue #1012)
+==================================
+
+The hook used to fire on the exact `prior_count == 1` boundary, so a session that
+kept replaying the same failure got exactly one advisory and then silence — and
+that silence is indistinguishable, to the model reading the transcript, from the
+loop having been noticed and accepted. The measured failure mode this hook exists
+for is precisely the long run (the poll-loop family recurred 6x in one session,
+5 of them after the first correction signal was already in-transcript), i.e. the
+occurrences the boundary suppressed. The advisory now repeats from the 2nd
+occurrence onward and names the count, so the signal gets stronger rather than
+vanishing exactly where the loop is worst.
 
 Failure detection
 ================
@@ -62,11 +75,17 @@ from _paths import resolve_cache_file  # type: ignore[import-not-found]  # noqa:
 from _state_lock import state_lock  # type: ignore[import-not-found]  # noqa: E402
 
 
-_ADVISORY_PREFIX = (
+# `{n}` is the running occurrence number (2, 3, 4, …) — issue #1012 replaced the
+# fixed "2회째" wording when the advisory stopped firing only on the 2nd failure.
+_ADVISORY_PREFIX_TMPL = (
     "[second-failure-advisory] "
-    "동일한 오류 패턴으로 세션 내 2회째 실패가 감지되었습니다. "
+    "동일한 오류 패턴으로 세션 내 {n}회째 실패가 감지되었습니다. "
     "원인 분석 없이 즉시 재시도하는 루프가 될 수 있습니다. "
 )
+
+# The advisory starts on the 2nd occurrence of a pair and repeats for every
+# occurrence after it (issue #1012).
+_ADVISORY_FROM_OCCURRENCE = 2
 
 _STATE_SCHEMA_VERSION = 1
 _MAX_SIGNATURE_LEN = 4_096
@@ -252,8 +271,8 @@ def _save_state(path: str, state: dict[str, Any]) -> bool:
         return False
 
 
-def _emit_advisory(tool_name: str, signature: str, reference: str) -> None:
-    """Emit the advisory once, on the 2nd failure.
+def _emit_advisory(tool_name: str, signature: str, reference: str, occurrence: int) -> None:
+    """Emit the advisory for the `occurrence`-th failure of this pair (>= 2).
 
     Written as `hookSpecificOutput.additionalContext` (DESIGN.md, PostToolUse
     corrective emissions), mirroring `builtin-task-postuse`. A PostToolUse hook
@@ -261,7 +280,8 @@ def _emit_advisory(tool_name: str, signature: str, reference: str) -> None:
     the stderr form would leave the retry loop uncorrected, which is the one
     thing this hook exists to do.
     """
-    message = f"{_ADVISORY_PREFIX}{tool_name} 실패 패턴 2회째 재현 중입니다. "
+    message = _ADVISORY_PREFIX_TMPL.format(n=occurrence)
+    message += f"{tool_name} 실패 패턴 {occurrence}회째 재현 중입니다. "
     message += f"signature={signature[:12]}"
     if reference:
         message += (
@@ -315,11 +335,14 @@ def main() -> int:
     path = _state_path(session_id)
     pair_key = f"{tool_name}\0{signature}"
 
-    # The whole read-modify-write is serialized (issue #951). The advisory
-    # fires on the exact `prior_count == 1` boundary, so two processes that
-    # read the same count both write count+1 and both cross that boundary —
-    # the duplicate fire recorded as unverified on #950. Persisting inside the
-    # lock is what makes each process observe the other's increment.
+    # The whole read-modify-write is serialized (issue #951). Two processes that
+    # read the same count both write count+1, so without the lock the same
+    # occurrence number is reported twice and one increment is lost — the
+    # duplicate fire recorded as unverified on #950. Persisting inside the lock
+    # is what makes each process observe the other's increment. (Issue #1012
+    # widened the fire condition from the `prior_count == 1` boundary to every
+    # occurrence >= 2; the lost increment still mis-numbers the advisory, so the
+    # lock is still load-bearing.)
     with state_lock(path):
         state = _load_state(path)
         failures = state.get("failures")
@@ -341,8 +364,9 @@ def main() -> int:
     if not saved:
         return 0
 
-    if prior_count == 1:
-        _emit_advisory(tool_name, signature, ref)
+    occurrence = prior_count + 1
+    if occurrence >= _ADVISORY_FROM_OCCURRENCE:
+        _emit_advisory(tool_name, signature, ref, occurrence)
     return 0
 
 

--- a/hooks/postuse-correction/second-failure-advisory/spec.md
+++ b/hooks/postuse-correction/second-failure-advisory/spec.md
@@ -4,7 +4,7 @@ Supported hosts: all
 
 `hooks/second-failure-advisory` is a PostToolUse 보조 훅입니다. 동일
 `tool_name + error_signature` 조합이 같은 세션에서 반복 실패했을 때
-2회째 실패에 한해 stdout의 `hookSpecificOutput.additionalContext`로 advisory를
+2회째 실패부터 stdout의 `hookSpecificOutput.additionalContext`로 advisory를
 출력해, 원인 분석 없이 동일 실패를 무한 재시도하는 패턴을 줄입니다.
 
 ## Why this exists
@@ -60,19 +60,32 @@ Supported hosts: all
 ## Counting semantics — 세션 누적, 연속 아님
 
 카운터는 `(tool_name, signature)` 쌍별 **세션 누적** 값입니다. 중간에 성공이나
-다른 실패가 끼어들어도 카운터는 리셋되지 않으며, 같은 쌍의 2회째 실패에서
-advisory가 발화합니다. 이는 issue #944가 지정한 동작("`(tool_name,
-error_signature)`를 세션 스코프로 카운트하고 2회차에 advisory")입니다.
+다른 실패가 끼어들어도 카운터는 리셋되지 않으며, 같은 쌍의 2회째 실패부터
+advisory가 발화합니다. issue #944가 지정한 동작은 "`(tool_name,
+error_signature)`를 세션 스코프로 카운트하고 2회차에 advisory"였습니다.
 
-"연속 실패"가 아니라 "세션 내 2회째 동일 실패"가 발화 조건이므로, 메시지와
+"연속 실패"가 아니라 "세션 내 N회째 동일 실패"가 발화 조건이므로, 메시지와
 문서 모두 연속(consecutive)이라는 표현을 쓰지 않습니다.
+
+## 3회째 이상도 계속 advisory (issue #1012)
+
+이전에는 `prior_count == 1` 경계에서만 발화했으므로, 같은 실패를 계속 반복하는
+세션은 advisory를 **정확히 한 번** 받고 그 뒤로는 침묵했습니다. transcript를
+읽는 모델 입장에서 그 침묵은 "루프를 인지하고 수용했다"와 구분되지 않습니다.
+이 훅이 존재하는 근거가 된 실측 패턴 자체가 긴 반복(poll-loop 계열이 한 세션에서
+6회 재발, 그중 5회는 첫 교정 신호가 이미 transcript에 있은 뒤)이므로, 경계가
+잘라낸 구간이 정확히 가장 필요한 구간이었습니다.
+
+이제 2회째부터 매 회 발화하며 메시지에 회차 번호(`{n}회째`)를 싣습니다. 신호가
+가장 나쁜 구간에서 사라지는 대신 누적됩니다. 1회째 무음은 그대로입니다 — 한 번의
+실패는 아직 루프가 아닙니다.
 
 ## Output behavior
 
-다음 경우에만 advisory를 출력합니다.
+다음 경우에 advisory를 출력합니다.
 
-- 동일 `session_id` 기준 2번째 실패 (`(tool_name, signature)` 조합)
-- 이전 카운트가 1일 때
+- 동일 `session_id` 기준 2회째 이상의 실패 (`(tool_name, signature)` 조합)
+- 즉 이전 카운트가 1 이상일 때 (`occurrence = prior_count + 1 >= 2`)
 
 상태 저장(`os.replace` 기반 원자적 교체)이 성공한 뒤에만 advisory를
 출력합니다. 저장에 실패하면 카운터가 남지 않아 같은 advisory가 다음 실패에서
@@ -81,9 +94,11 @@ error_signature)`를 세션 스코프로 카운트하고 2회차에 advisory")�
 원자적인 것은 교체(rename)뿐이며, 읽기-증가-쓰기-발화 전체는 프로세스 간에
 직렬화되지 않습니다. 한 세션에서 도구 호출이 병렬로 끝나면 PostToolUse는
 호출마다 별도 프로세스로 돌므로, 저장된 count가 1일 때 두 프로세스가 함께
-1을 읽어 각자 2를 쓰고 **둘 다 advisory를 낼 수 있습니다**. 반대로 서로 다른
-쌍의 동시 실패는 한쪽 증가분을 덮어써 advisory가 한 번 늦어질 수 있습니다.
-"3회째 이상 무음"은 순차 실행 기준의 계약이며, 이 창에서는 지켜지지 않습니다.
+1을 읽어 각자 2를 쓰고 **둘 다 같은 회차 번호로 advisory를 낼 수 있습니다**.
+반대로 서로 다른 쌍의 동시 실패는 한쪽 증가분을 덮어써 advisory가 한 번
+늦어질 수 있습니다. 회차 번호가 실제 실패 횟수와 일치한다는 것은 순차 실행
+기준의 계약이며, 이 창에서는 지켜지지 않습니다 (issue #1012 이전에는 같은
+창에서 "3회째 이상 무음" 계약이 깨졌습니다).
 
 lock은 두지 않습니다. 피해가 advisory 한 줄 중복 또는 한 박자 지연에 그치고
 이 훅은 차단하지 않기 때문이며, 무엇보다 동일한 읽기-수정-쓰기 + `os.replace`
@@ -98,8 +113,10 @@ exit 0인 PostToolUse 훅의 `stderr`는 디버그 로그로만 가고 모델에
 일어나지 않습니다.
 
 ```json
-{"continue": true, "hookSpecificOutput": {"hookEventName": "PostToolUse", "additionalContext": "[second-failure-advisory] 동일한 오류 패턴으로 세션 내 2회째 실패가 감지되었습니다. … signature=<sig_prefix> Reference: <path?> — …"}}
+{"continue": true, "hookSpecificOutput": {"hookEventName": "PostToolUse", "additionalContext": "[second-failure-advisory] 동일한 오류 패턴으로 세션 내 <n>회째 실패가 감지되었습니다. … signature=<sig_prefix> Reference: <path?> — …"}}
 ```
+
+`<n>`은 해당 `(tool_name, signature)` 쌍의 세션 누적 회차(2, 3, 4, …)입니다.
 
 `reference`는 실패 텍스트의 `Reference:` label, `hooks/...` 또는 `*spec.md`
 경로에서 먼저 추출하고, 없으면 `tool_input.file_path/path/target`을 사용합니다.
@@ -164,9 +181,9 @@ python3 -m pytest tests/test_hook_state_concurrency.py
 
 필수 커버:
 
-- 1회 실패: advisory 없음
+- 1회 실패: advisory 없음 (양방향 control — "항상 발화"로 퇴화하면 이 케이스가 잡음)
 - 2회 실패(동일 signature): advisory 출력
-- 3회째 이상(동일 쌍): 추가 advisory 없음
+- 3, 4, 5회째(동일 쌍): 계속 advisory 출력, 메시지에 회차 번호 포함 (issue #1012)
 - 동일 시그니처에서 tool_name이 다르면 advisory 없음
 - 경로/해시/타임스탬프만 바뀐 2회 실패도 advisory 출력
 - 사이에 성공/다른 실패가 끼어도 같은 쌍의 2회째에 advisory 출력
@@ -175,4 +192,4 @@ python3 -m pytest tests/test_hook_state_concurrency.py
 - 실패 텍스트의 `Reference:` 경로가 advisory와 재진술 지시에 포함됨
 - 비실패/비정상 입력은 fail-open
 - 두 프로세스 동시 실행: 잠금 없이는 증분 유실, 잠금 하에서는 카운트 1→2→3
-  과 advisory 1회 (`tests/test_hook_state_concurrency.py`)
+  과 advisory 2회(2회째·3회째) (`tests/test_hook_state_concurrency.py`)

--- a/hooks/preflight-gate/foreground-poll-loop-guard/impl.py
+++ b/hooks/preflight-gate/foreground-poll-loop-guard/impl.py
@@ -52,9 +52,43 @@ Exit 0 (pass) otherwise: run_in_background=true, short bounded loops (< ~90s),
 no-sleep loops, leading `sleep` without a loop (the runtime already handles that),
 non-Bash tools.
 
+## Read-gate escalation (issue #1012)
+
+A block that is answered with the same loop shape is a block that was not read.
+From the (N+1)-th block of THIS guard in the same session (N =
+`_READ_GATE_AFTER_BLOCKS`), the block message escalates: it names this guard's
+own `spec.md` by absolute path and states that the poll-loop retry stays denied
+until that file has been Read in this session.
+
+Three properties keep the escalation from pointing at nothing or deadlocking:
+
+  1. The referent is this guard's own `spec.md`, not the 132-byte
+     `docs/hook/…md` redirect stub it used to name.
+  2. The path is resolved from the PACKAGE ROOT (`__file__`), so it is absolute
+     and resolves identically whether or not the agent's cwd is a praxis
+     checkout — a bare repo-relative path resolves nowhere else, so the agent
+     that DID read the spec would still be denied.
+  3. If that path does not exist, the escalation FAILS OPEN to the base block.
+     A gate whose own satisfaction condition is unreachable must not add a
+     requirement.
+
+The prior-block count comes from the fire ledger's
+`count_session_fires(hook, session_id, decision="block")` — the dispatcher
+already writes a RICH `(session_id, hook, decision)` row per member per Bash
+call, so no second counter is introduced. The read-set is the one
+`pre-edit-md-escape-advisory` already maintains (its PostToolUse(Read) leg
+records EVERY `.md` Read), so no second read-history is introduced either.
+
+The escalation only ever changes the MESSAGE of a command this guard was
+already going to block — it can never turn a pass into a deny.
+
 ## Env vars
 
 - `PRAXIS_HOOK_BYPASS_POLL_LOOP_GUARD` — set to any non-empty value → bypass (exit 0).
+- `PRAXIS_POLL_LOOP_GUARD_SPEC` — override the spec path the read-gate points at
+  (tests; an unresolvable value exercises the fail-open escape).
+- `PRAXIS_POLL_LOOP_READ_GATE_AFTER` — prior blocks required before the read-gate
+  escalates (default 2 → the 3rd block escalates). Unparseable → default.
 
 ## Fail-open
 
@@ -69,15 +103,31 @@ import re
 import sys
 from pathlib import Path as _Path
 
-_sys_lib = str(_Path(__file__).resolve().parent.parent.parent / "_lib")
+_HERE = _Path(__file__).resolve()
+# <root>/hooks/<role>/<name>/impl.py — parents[2] is hooks/. Anchoring on
+# __file__ (never on cwd) is what makes the paths below resolve the same in a
+# checkout and in an installed plugin; it is the same package-root rule
+# `_lib/_fire_ledger._checkout_root` states for the ledger.
+_HOOKS_DIR = _HERE.parents[2]
+
+_sys_lib = str(_HOOKS_DIR / "_lib")
 if _sys_lib not in sys.path:
     sys.path.insert(0, _sys_lib)
+from _fire_ledger import count_session_fires  # type: ignore[import-not-found]  # noqa: E402
 from _hook_runtime import fail_open  # type: ignore[import-not-found]  # noqa: E402
 from _hook_utils import safe_tokenize  # type: ignore[import-not-found]  # noqa: E402
 from block_message import emit_block  # type: ignore[import-not-found]  # noqa: E402
 
+_HOOK_NAME = "foreground-poll-loop-guard"  # manifest `name` — the fire-ledger key
 _FOREGROUND_CEILING_S = 120  # Bash default timeout, seconds
 _SAFE_MARGIN_S = 100  # block when worst-case can approach the ceiling
+_READ_GATE_AFTER_BLOCKS = 2  # prior session blocks before the read-gate escalates
+# The read-set owner: its PostToolUse(Read) leg records every `.md` Read of the
+# session, spec.md included, so the escalation reuses that history instead of
+# standing up a second one.
+_MD_READ_HISTORY_IMPL = (
+    _HOOKS_DIR / "postuse-correction" / "pre-edit-md-escape-advisory" / "impl.py"
+)
 
 _LOOP_KEYWORDS = {"for", "while", "until"}
 # Separator tokens after which the next token sits in command position (bash
@@ -303,6 +353,85 @@ def _detect(command: str) -> str | None:
     return None
 
 
+def _reference_path() -> str:
+    """Absolute path to this guard's own spec.md (issue #1012).
+
+    Package-root-anchored, never cwd-relative: the reference is both what the
+    block message prints AND the key the read-gate looks up in the session's
+    read-set, so a repo-relative string would name a file that does not exist
+    when the agent is working outside the praxis checkout — the agent that
+    followed the reference would still be denied. `PRAXIS_POLL_LOOP_GUARD_SPEC`
+    overrides it so tests can point the gate at a path that does not resolve.
+    """
+    override = os.environ.get("PRAXIS_POLL_LOOP_GUARD_SPEC", "").strip()
+    if override:
+        return os.path.abspath(os.path.expanduser(override))
+    return str(_HERE.parent / "spec.md")
+
+
+def _read_gate_threshold() -> int:
+    """Prior session blocks required before the read-gate escalates."""
+    raw = os.environ.get("PRAXIS_POLL_LOOP_READ_GATE_AFTER", "").strip()
+    try:
+        return max(int(raw), 0) if raw else _READ_GATE_AFTER_BLOCKS
+    except ValueError:
+        return _READ_GATE_AFTER_BLOCKS  # unparseable → default, never a crash
+
+
+def _spec_read_in_session(session_id: str, spec_path: str) -> bool:
+    """True when `spec_path` was Read in this session (issue #1012).
+
+    Reuses the read-set `pre-edit-md-escape-advisory` already maintains — its
+    PostToolUse(Read) leg records EVERY `.md` path Read in the session, so
+    spec.md is already in there and a second read-history mechanism would only
+    be a second thing to drift. Loaded by file location under a unique module
+    name (that sibling hook's module is also called `impl`) and only from the
+    escalation branch, which runs solely on a command this guard is already
+    blocking — the hot Bash path never pays for the import.
+
+    Any failure returns True (= "treat as read"), so a broken lookup relaxes the
+    gate instead of denying on the strength of state it could not read.
+    """
+    try:
+        import importlib.util
+
+        spec = importlib.util.spec_from_file_location(
+            "_praxis_poll_loop_md_read_history", _MD_READ_HISTORY_IMPL
+        )
+        if spec is None or spec.loader is None:
+            return True
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
+        history_path = module.resolve_history_path(session_id)
+        return module.normalize_path(spec_path) in module.get_read_set(history_path)
+    except Exception:
+        return True
+
+
+def _read_gate_engaged(payload: dict, reference: str) -> int | None:
+    """Prior-block count when the read-gate should escalate, else None.
+
+    Returns None (no escalation) when: the payload carries no session_id, the
+    reference does not resolve on disk (the mandatory fail-open escape — a gate
+    that cannot find its own spec must not add a requirement nobody can
+    satisfy), the session has not been blocked enough times yet, or the spec has
+    already been Read this session.
+    """
+    session_id = payload.get("session_id")
+    if not isinstance(session_id, str) or not session_id:
+        return None
+    if not os.path.isfile(reference):
+        return None
+    # The dispatcher records a member's fire AFTER running its main(), so this
+    # is the count of PRIOR blocks this session, not including the in-flight one.
+    prior_blocks = count_session_fires(_HOOK_NAME, session_id, decision="block")
+    if prior_blocks < _read_gate_threshold():
+        return None
+    if _spec_read_in_session(session_id, reference):
+        return None
+    return prior_blocks
+
+
 @fail_open
 def main() -> int:
     if os.environ.get("PRAXIS_HOOK_BYPASS_POLL_LOOP_GUARD"):
@@ -329,19 +458,40 @@ def main() -> int:
     if reason is None:
         return 0
 
+    # Issue #1012: the referent used to be `docs/hook/foreground-poll-loop-guard.md`,
+    # a 132-byte redirect stub whose entire content is "Moved to <spec.md>". The
+    # gate now names the spec itself, by absolute package-root-anchored path.
+    reference = _reference_path()
+    correct_path = (
+        "use a native async-wait primitive: run_in_background: true "
+        "(Claude is re-invoked on exit — no polling loop needed), a "
+        "Monitor until-loop, `aws cloudformation wait`, `gh run watch` / "
+        "`gh pr checks --watch` / `kubectl wait`; a short bounded "
+        "foreground poll (worst-case < 90s) is fine"
+    )
+    why = (
+        f"{reason} — the foreground call dies with SIGTERM (Exit 143) "
+        "mid-poll even when the awaited async op succeeds"
+    )
+
+    prior_blocks = _read_gate_engaged(payload, reference)
+    if prior_blocks is not None:
+        why = (
+            f"{why}. READ-GATE: this guard has already blocked {prior_blocks}x in "
+            "this session and the same loop shape came back, so the block message "
+            "is not being consumed"
+        )
+        correct_path = (
+            f"Read {reference} FIRST — a poll-loop retry stays denied until that "
+            f"Read is recorded in this session. Then {correct_path}"
+        )
+
     emit_block(
         rule_name="foreground poll-loop guard",
-        why=f"{reason} — the foreground call dies with SIGTERM (Exit 143) "
-            "mid-poll even when the awaited async op succeeds",
-        correct_path=(
-            "use a native async-wait primitive: run_in_background: true "
-            "(Claude is re-invoked on exit — no polling loop needed), a "
-            "Monitor until-loop, `aws cloudformation wait`, `gh run watch` / "
-            "`gh pr checks --watch` / `kubectl wait`; a short bounded "
-            "foreground poll (worst-case < 90s) is fine"
-        ),
+        why=why,
+        correct_path=correct_path,
         bypass_env="PRAXIS_HOOK_BYPASS_POLL_LOOP_GUARD",
-        reference="docs/hook/foreground-poll-loop-guard.md",
+        reference=reference,
     )
     return 2
 

--- a/hooks/preflight-gate/foreground-poll-loop-guard/spec.md
+++ b/hooks/preflight-gate/foreground-poll-loop-guard/spec.md
@@ -80,11 +80,84 @@ The block message names the alternatives so the caller can self-correct:
 `run_in_background: true`, Monitor with an until-loop, `aws cloudformation wait`,
 `gh run watch` / `gh pr checks --watch` / `kubectl wait`.
 
+`Reference:` is this file, named by ABSOLUTE path resolved from the package root
+(`impl.py`'s own `__file__`), never a cwd-relative string. Until issue #1012 it
+named `docs/hook/foreground-poll-loop-guard.md` — a 132-byte redirect stub whose
+entire content is "Moved to …" — and named it repo-relatively, so it resolved to
+nothing whenever the agent was working outside a praxis checkout.
+
+## Read-gate escalation (issue #1012)
+
+A block answered with the same loop shape is a block that was not read. From the
+(N+1)-th block of THIS guard in the same session (N = 2 by default) the message
+escalates: `Why:` gains a `READ-GATE:` clause naming the running block count, and
+`Correct path:` leads with "Read `<abs spec.md>` FIRST — a poll-loop retry stays
+denied until that Read is recorded in this session".
+
+| Situation (all on a command the guard already blocks) | Message |
+| --- | --- |
+| 1st / 2nd block of the session | base block |
+| 3rd+ block, this spec not Read this session | **escalated** (`READ-GATE`) |
+| 3rd+ block, this spec Read this session | base block |
+| 3rd+ block, some *other* `.md` Read | **escalated** (read-set is path-keyed) |
+| 3rd+ block, reference path does not resolve | base block (**fail-open**) |
+| payload carries no `session_id` | base block |
+| `run_in_background: true` while escalation is armed | pass (exit 0) |
+
+Two mechanisms are REUSED rather than rebuilt — a second counter or a second
+read-history would only be a second thing to drift:
+
+- **Prior-block count** — `_lib/_fire_ledger.count_session_fires(hook,
+  session_id, decision="block")`. The dispatcher already writes one RICH
+  `(session_id, hook, decision)` row per group member per Bash call, and it
+  writes it AFTER the member's `main()`, so the value read here is the count of
+  PRIOR blocks, not including the in-flight one. Standalone (non-dispatched)
+  invocation produces only COARSE rows (`session_id: ""`), which
+  `count_session_fires` filters out — so the escalation is a dispatched-runtime
+  behavior, and a lone `python3 impl.py` probe never sees it.
+- **Read-set** — the session history `pre-edit-md-escape-advisory` maintains. Its
+  PostToolUse(Read) leg records EVERY `.md` path Read in the session, spec.md
+  included. That module is loaded lazily by file location (its module is also
+  named `impl`) and only from the escalation branch, which runs solely on a
+  command that is already being blocked, so the hot Bash path never pays for it.
+
+Three properties keep the escalation from pointing at nothing or deadlocking:
+the referent is this spec (not the redirect stub); the path is absolute and
+package-root-anchored, so the agent that DID follow the reference is recognized
+wherever it is working; and an unresolvable referent fails OPEN to the base
+block. A gate that cannot find its own spec must not add a requirement nobody
+can satisfy.
+
+The escalation only ever rewrites the MESSAGE of a command this guard was
+already going to block. It can never turn a pass into a deny.
+
 ## Env vars
 
 - `PRAXIS_HOOK_BYPASS_POLL_LOOP_GUARD` — set to any non-empty value → bypass (exit 0).
+- `PRAXIS_POLL_LOOP_GUARD_SPEC` — override the path the read-gate points at
+  (tests; an unresolvable value exercises the fail-open escape).
+- `PRAXIS_POLL_LOOP_READ_GATE_AFTER` — prior session blocks required before the
+  read-gate escalates (default `2` → the 3rd block escalates). Unparseable →
+  default.
 
 ## Fail-open
 
 Malformed stdin JSON, non-Bash tool, empty command, unparseable count/sleep →
-exit 0 (pass). The guard never blocks on infrastructure error.
+exit 0 (pass). The guard never blocks on infrastructure error. Within the block
+path, the read-gate escalation additionally fails open (base block, no Read
+demanded) on a missing `session_id`, an unresolvable reference path, or any
+error reading the fire ledger or the read-set.
+
+## Tests
+
+```bash
+bash tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
+```
+
+Read-gate coverage is deliberately two-directional — a one-sided fixture would
+let the opposite error in. Fire direction: nth un-Read block escalates; an
+unrelated `.md` Read does not satisfy it; a count written by the REAL dispatcher
+escalates (the case that fails if `record_group_fires`' row shape ever drifts
+from `count_session_fires`' filter). Silence direction: below threshold, after a
+Read of this spec, unresolvable reference, missing `session_id`, and
+`run_in_background: true` while armed.

--- a/tests/hooks/postuse-correction/test_second_failure_advisory.sh
+++ b/tests/hooks/postuse-correction/test_second_failure_advisory.sh
@@ -15,7 +15,8 @@
 #   9) 상태 저장 실패 시 advisory 무음
 #  10) stdout/output만 있는 성공 응답은 반복돼도 무음
 #  11) 실패 텍스트의 Reference 경로가 advisory에 포함
-#  12) 3회째 이상은 추가 advisory 없음
+#  12) 3회째 이상도 계속 advisory (회차 번호 포함) — issue #1012
+#  12b) 첫 회는 여전히 무음 (반대 방향 control)
 #  13) interrupted 응답은 실패로 판정
 #
 # Run:
@@ -408,23 +409,49 @@ else
 fi
 
 # ---------------------------------------------------------------------------
-# Case 12: the advisory fires once — the 3rd+ failure adds nothing
+# Case 12: occurrences 3..N keep advising, numbered (issue #1012)
+#
+# The old contract stopped at the `prior_count == 1` boundary, so a session that
+# kept replaying the same failure went silent exactly where the loop was worst.
+# Own state file (`c12b`) — case 8 above already writes `c12.json`, and sharing
+# it made the occurrence number here depend on an unrelated case.
 # ---------------------------------------------------------------------------
-echo "=== case 12: third failure => no further advisory ==="
-STATE12="$TMP_DIR/c12.json"
+echo "=== case 12: occurrences 3..N keep advising, numbered ==="
+STATE12B="$TMP_DIR/c12b.json"
 payload12="$(make_payload Bash sess-944-third 1)"
-pipe_hook "$payload12" "$STATE12" >/dev/null 2>/dev/null
-pipe_hook "$payload12" "$STATE12" >/dev/null 2>/dev/null
+pipe_hook "$payload12" "$STATE12B" >/dev/null 2>/dev/null   # 1st — silent
+pipe_hook "$payload12" "$STATE12B" >/dev/null 2>/dev/null   # 2nd — advises
+
+for occurrence in 3 4 5; do
+  out_file="$(mktemp)" err_file="$(mktemp)"
+  pipe_hook "$payload12" "$STATE12B" >"$out_file" 2>"$err_file"
+  rc=$?
+  out=$(cat "$out_file"); err=$(cat "$err_file")
+  rm -f "$out_file" "$err_file"
+
+  if [ "$rc" -eq 0 ] && [ -z "$err" ] && [ -n "$out" ] \
+    && assert_match "${occurrence}회째" "$out"; then
+    assert_pass "12) occurrence ${occurrence} advises and names its count"
+  else
+    assert_fail "12) occurrence ${occurrence} advises and names its count" \
+      "rc=$rc out=[$out] err=[$err]"
+  fi
+done
+
+# Control for the same boundary in the other direction: occurrence 1 of a FRESH
+# pair must still be silent. Without it, "always advise" would pass case 12 too.
+echo "=== case 12b: first occurrence still silent (control) ==="
+STATE12C="$TMP_DIR/c12c.json"
 out_file="$(mktemp)" err_file="$(mktemp)"
-pipe_hook "$payload12" "$STATE12" >"$out_file" 2>"$err_file"
+pipe_hook "$(make_payload Bash sess-944-first-only 1)" "$STATE12C" >"$out_file" 2>"$err_file"
 rc=$?
 out=$(cat "$out_file"); err=$(cat "$err_file")
 rm -f "$out_file" "$err_file"
 
 if [ "$rc" -eq 0 ] && [ -z "$out" ] && [ -z "$err" ]; then
-  assert_pass "12) third failure emits no further advisory"
+  assert_pass "12b) first occurrence stays silent"
 else
-  assert_fail "12) third failure emits no further advisory" "rc=$rc out=[$out] err=[$err]"
+  assert_fail "12b) first occurrence stays silent" "rc=$rc out=[$out] err=[$err]"
 fi
 
 # ---------------------------------------------------------------------------

--- a/tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
+++ b/tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
@@ -183,6 +183,155 @@ err=$(echo '{"tool_name": "Bash", "tool_input": {"command": "while true; do slee
 if [ "$rc" -eq 0 ]; then PASS=$((PASS + 1)); echo "ok    [env-bypass]"; else
   echo "FAIL  [env-bypass] expected exit 0, got $rc"; FAIL=$((FAIL + 1)); FAILED_NAMES+=("env-bypass"); fi
 
+# ---- read-gate escalation (issue #1012) ---------------------------------------
+#
+# From the (N+1)-th block of this guard in one session the message escalates to
+# "Read <spec.md> first". The prior-block count comes from the fire ledger's RICH
+# rows, so the fixtures seed that ledger in the exact shape record_group_fires
+# writes — and one case drives the REAL dispatcher, which is the only thing that
+# proves writer and reader still agree on that shape.
+#
+# Both directions are pinned: the gate must fire on the nth un-Read block, and it
+# must stay silent below the threshold, after a Read, and when its own referent
+# does not resolve (the mandatory fail-open escape — a gate that cannot find its
+# spec must not demand it be read).
+
+RG_DIR="$(mktemp -d)" || { echo "FATAL: mktemp -d failed" >&2; exit 1; }
+trap 'rm -rf "$RG_DIR"' EXIT
+RG_SPEC="$REPO_ROOT/hooks/preflight-gate/foreground-poll-loop-guard/spec.md"
+RG_MD_POST="$REPO_ROOT/hooks/postuse-correction/pre-edit-md-escape-advisory/impl.py"
+RG_LOOP='while true; do gh pr checks 7; sleep 20; done'
+
+# seed_blocks <ledger> <session_id> <count> — RICH block rows for this guard.
+seed_blocks() {
+  python3 - "$1" "$2" "$3" <<'PY'
+import json, sys
+ledger, sid, n = sys.argv[1], sys.argv[2], int(sys.argv[3])
+with open(ledger, "a", encoding="utf-8") as fh:
+    for _ in range(n):
+        fh.write(json.dumps({
+            "timestamp": "2026-01-01T00:00:00+00:00",
+            "session_id": sid,
+            "tool": "Bash",
+            "hook": "foreground-poll-loop-guard",
+            "role": "preflight-gate",
+            "decision": "block",
+            "granularity": "rich",
+        }) + "\n")
+PY
+}
+
+# rg_payload <session_id> [bg]
+rg_payload() {
+  python3 - "$1" "${2:-}" "$RG_LOOP" <<'PY'
+import json, sys
+sid, bg, cmd = sys.argv[1], sys.argv[2], sys.argv[3]
+tool_input = {"command": cmd}
+if bg == "bg":
+    tool_input["run_in_background"] = True
+print(json.dumps({"session_id": sid, "tool_name": "Bash", "tool_input": tool_input}))
+PY
+}
+
+# rg_run <ledger> <history> <payload> [spec_override] — stderr of one guard run.
+rg_run() {
+  local ledger="$1" history="$2" payload="$3" spec_override="${4:-}"
+  (
+    export PRAXIS_FIRE_TELEMETRY_FILE="$ledger"
+    export PRAXIS_MD_READ_HISTORY_FILE="$history"
+    if [ -n "$spec_override" ]; then export PRAXIS_POLL_LOOP_GUARD_SPEC="$spec_override"; fi
+    printf '%s' "$payload" | "$HOOK" 2>&1 >/dev/null
+  )
+}
+
+# rg_assert <name> <gate|nogate> <expected_rc> <actual_rc> <stderr>
+rg_assert() {
+  local name="$1" want="$2" want_rc="$3" rc="$4" err="$5"
+  if [ "$rc" -ne "$want_rc" ]; then
+    echo "FAIL  [$name] expected exit $want_rc, got $rc (stderr: ${err:-<empty>})"
+    FAIL=$((FAIL + 1)); FAILED_NAMES+=("$name"); return
+  fi
+  if [ "$want" = "gate" ] && ! echo "$err" | grep -q 'READ-GATE'; then
+    echo "FAIL  [$name] expected READ-GATE escalation, got: ${err:-<empty>}"
+    FAIL=$((FAIL + 1)); FAILED_NAMES+=("$name"); return
+  fi
+  if [ "$want" = "nogate" ] && echo "$err" | grep -q 'READ-GATE'; then
+    echo "FAIL  [$name] expected NO escalation, got: ${err:-<empty>}"
+    FAIL=$((FAIL + 1)); FAILED_NAMES+=("$name"); return
+  fi
+  PASS=$((PASS + 1)); echo "ok    [$name]"
+}
+
+# The block message must name the spec by ABSOLUTE path: it is both what the
+# agent is told to read and the key the read-set is looked up by, and a
+# repo-relative string resolves nowhere outside a praxis checkout.
+err=$(rg_run "$RG_DIR/ref.jsonl" "$RG_DIR/ref-history.json" "$(rg_payload ref-sess)"); rc=$?
+if [ "$rc" -eq 2 ] && echo "$err" | grep -qF "Reference: $RG_SPEC"; then
+  PASS=$((PASS + 1)); echo "ok    [readgate-reference-is-absolute-spec]"
+else
+  echo "FAIL  [readgate-reference-is-absolute-spec] got: ${err:-<empty>}"
+  FAIL=$((FAIL + 1)); FAILED_NAMES+=("readgate-reference-is-absolute-spec")
+fi
+
+# 1 prior block is below the default threshold of 2 → base block only.
+seed_blocks "$RG_DIR/below.jsonl" below-sess 1
+err=$(rg_run "$RG_DIR/below.jsonl" "$RG_DIR/below-history.json" "$(rg_payload below-sess)"); rc=$?
+rg_assert "readgate-below-threshold-silent" nogate 2 "$rc" "$err"
+
+# 2 prior blocks, spec never Read → the 3rd block escalates.
+seed_blocks "$RG_DIR/nth.jsonl" nth-sess 2
+err=$(rg_run "$RG_DIR/nth.jsonl" "$RG_DIR/nth-history.json" "$(rg_payload nth-sess)"); rc=$?
+rg_assert "readgate-nth-block-demands-read" gate 2 "$rc" "$err"
+
+# Same state, but the spec was Read this session (recorded by the read-set owner
+# pre-edit-md-escape-advisory, not by a second mechanism) → escalation released.
+seed_blocks "$RG_DIR/read.jsonl" read-sess 2
+printf '{"session_id":"read-sess","tool_name":"Read","tool_input":{"file_path":"%s"}}' "$RG_SPEC" \
+  | PRAXIS_MD_READ_HISTORY_FILE="$RG_DIR/read-history.json" python3 "$RG_MD_POST" post >/dev/null 2>&1
+err=$(rg_run "$RG_DIR/read.jsonl" "$RG_DIR/read-history.json" "$(rg_payload read-sess)"); rc=$?
+rg_assert "readgate-released-after-read" nogate 2 "$rc" "$err"
+
+# A Read of some OTHER .md must not satisfy the gate (the read-set is keyed by
+# path — without this the previous case would pass for the wrong reason).
+seed_blocks "$RG_DIR/other.jsonl" other-sess 2
+printf '{"session_id":"other-sess","tool_name":"Read","tool_input":{"file_path":"%s"}}' "$RG_DIR/unrelated.md" \
+  | PRAXIS_MD_READ_HISTORY_FILE="$RG_DIR/other-history.json" python3 "$RG_MD_POST" post >/dev/null 2>&1
+err=$(rg_run "$RG_DIR/other.jsonl" "$RG_DIR/other-history.json" "$(rg_payload other-sess)"); rc=$?
+rg_assert "readgate-other-md-read-does-not-satisfy" gate 2 "$rc" "$err"
+
+# Unresolvable referent → fail OPEN to the base block. Deadlock escape: the gate
+# must never demand a Read of a file that is not there.
+seed_blocks "$RG_DIR/noref.jsonl" noref-sess 2
+err=$(rg_run "$RG_DIR/noref.jsonl" "$RG_DIR/noref-history.json" "$(rg_payload noref-sess)" \
+  "$RG_DIR/absent/spec.md"); rc=$?
+rg_assert "readgate-unresolvable-reference-fails-open" nogate 2 "$rc" "$err"
+
+# No session_id → nothing to count against → base block.
+seed_blocks "$RG_DIR/nosid.jsonl" nosid-sess 2
+err=$(rg_run "$RG_DIR/nosid.jsonl" "$RG_DIR/nosid-history.json" \
+  '{"tool_name":"Bash","tool_input":{"command":"while true; do gh pr checks 7; sleep 20; done"}}'); rc=$?
+rg_assert "readgate-no-session-id-silent" nogate 2 "$rc" "$err"
+
+# The escalation only ever rewrites the message of a command that was already
+# being blocked — an armed gate must not deny a correct retry.
+seed_blocks "$RG_DIR/bg.jsonl" bg-sess 5
+err=$(rg_run "$RG_DIR/bg.jsonl" "$RG_DIR/bg-history.json" "$(rg_payload bg-sess bg)"); rc=$?
+rg_assert "readgate-armed-still-passes-background" nogate 0 "$rc" "$err"
+
+# End to end: the REAL dispatcher writes the rows the read-gate reads. This is
+# the case that fails if record_group_fires' record shape ever drifts from
+# count_session_fires' filter.
+(
+  export PRAXIS_FIRE_TELEMETRY_FILE="$RG_DIR/e2e.jsonl"
+  export PRAXIS_MD_READ_HISTORY_FILE="$RG_DIR/e2e-history.json"
+  export PRAXIS_HOME="$RG_DIR/e2e-home"   # contain every hook's state writes
+  for _ in 1 2; do
+    rg_payload e2e-sess | python3 "$REPO_ROOT/hooks/_lib/_dispatch.py" PreToolUse Bash >/dev/null 2>&1
+  done
+)
+err=$(rg_run "$RG_DIR/e2e.jsonl" "$RG_DIR/e2e-history.json" "$(rg_payload e2e-sess)"); rc=$?
+rg_assert "readgate-dispatcher-written-count-escalates" gate 2 "$rc" "$err"
+
 # ---- summary ------------------------------------------------------------------
 echo ""
 echo "passed: $PASS  failed: $FAIL"

--- a/tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
+++ b/tests/hooks/preflight-gate/test_foreground_poll_loop_guard.sh
@@ -240,7 +240,10 @@ rg_run() {
     export PRAXIS_FIRE_TELEMETRY_FILE="$ledger"
     export PRAXIS_MD_READ_HISTORY_FILE="$history"
     if [ -n "$spec_override" ]; then export PRAXIS_POLL_LOOP_GUARD_SPEC="$spec_override"; fi
-    printf '%s' "$payload" | "$HOOK" 2>&1 >/dev/null
+    # Braces, not a bare `2>&1 >/dev/null`: both orderings capture stderr and
+    # drop stdout, but shellcheck reads the bare form as a likely mistake
+    # (SC2069) and the CI shellcheck job is blocking.
+    { printf '%s' "$payload" | "$HOOK" >/dev/null; } 2>&1
   )
 }
 

--- a/tests/test_hook_state_concurrency.py
+++ b/tests/test_hook_state_concurrency.py
@@ -251,15 +251,17 @@ def _counts(state_file: Path) -> list[int]:
         # Both children read count=1 and both write 2, so the third failure is
         # never recorded. What reaches the model then splits two ways, and the
         # split is a scheduling detail rather than something to pin: usually
-        # both children cross the `prior_count == 1` boundary and the advisory
-        # fires twice (the duplicate recorded as unverified on #950), but the
-        # two also share one `<path>.tmp` staging name, so one child's
-        # `os.replace` can find it already renamed away, fail, and suppress its
-        # own advisory. The lost increment is the invariant violation common to
-        # both, and it is deterministic.
+        # both children advise as "2회째" (the duplicate recorded as unverified
+        # on #950), but the two also share one `<path>.tmp` staging name, so one
+        # child's `os.replace` can find it already renamed away, fail, and
+        # suppress its own advisory. The lost increment is the invariant
+        # violation common to both, and it is deterministic — it is also what
+        # the count assertion below, not the advisory count, discriminates on.
         ("nolock", 2, (1, 2)),
-        # Serialized: 1 -> 2 (advises) -> 3 (silent).
-        ("lock", 3, (1,)),
+        # Serialized: 1 -> 2 (advises) -> 3 (advises too, since #1012 widened
+        # the fire condition from the `prior_count == 1` boundary to every
+        # occurrence >= 2).
+        ("lock", 3, (2,)),
     ],
 )
 def test_second_failure_advisory_race(

